### PR TITLE
fix(idempotency): make idempotent_function decorator thread safe

### LINKED ISSUE #1718: [Thread-safety] Concurrent instantiation of Boto3 clients
### Expected Behaviour

The decorator `idempotent_function` should be thread-safe. In particular, consumers of the library should not have to worry about threading issues in the underlying library. 

In particular, the code snippet provided below should run without errors. 

### Current Behaviour

AWS Powertools instantiates boto3 client objects concurrently. While boto3 client objects are thread-safe, their instantiation is not. 
Hence, we see exceptions like the following:

> Traceback (most recent call last):
  File "ROOTidempotency_problem\venv\lib\site-packages\aws_lambda_powertools\utilities\idempotency\base.py", line 106, in _process_idempotency
    self.persistence_store.save_inprogress(
  File "ROOTidempotency_problem\venv\lib\site-packages\aws_lambda_powertools\utilities\idempotency\persistence\base.py", line 370, in save_inprogress
    self._put_record(data_record=data_record)
  File "ROOTidempotency_problem\venv\lib\site-packages\aws_lambda_powertools\utilities\idempotency\persistence\dynamodb.py", line 218, in _put_record
    except self.table.meta.client.exceptions.ConditionalCheckFailedException:
  File "ROOTidempotency_problem\venv\lib\site-packages\aws_lambda_powertools\utilities\idempotency\persistence\dynamodb.py", line 108, in table
    ddb_resource = self._boto3_session.resource("dynamodb", config=self._boto_config)
  File "ROOTidempotency_problem\venv\lib\site-packages\boto3\session.py", line 446, in resource
    client = self.client(
  File "ROOTidempotency_problem\venv\lib\site-packages\boto3\session.py", line 299, in client
    return self._session.create_client(
  File "ROOTidempotency_problem\venv\lib\site-packages\botocore\session.py", line 953, in create_client
    endpoint_resolver = self._get_internal_component('endpoint_resolver')
  File "ROOTidempotency_problem\venv\lib\site-packages\botocore\session.py", line 812, in _get_internal_component
    return self._internal_components.get_component(name)
  File "ROOTidempotency_problem\venv\lib\site-packages\botocore\session.py", line 1112, in get_component
    del self._deferred[name]
KeyError: 'endpoint_resolver'

### Code snippet

```python
# Two threads, reliably reproducing the issue with 1.31.1 as well as the latest version (2.2.0): 

import functools
import inspect
import threading
import time
from dataclasses import dataclass
from enum import Enum
from botocore.config import Config

from threading import Thread
from typing import Callable

from aws_lambda_powertools.utilities.idempotency import IdempotencyConfig, idempotent_function
from aws_lambda_powertools.utilities.idempotency import DynamoDBPersistenceLayer as OriginalDynamoDBPersistenceLayer

idempotentTableName: str = 'trading_idempotency'


class DynamoDBPersistenceLayer(OriginalDynamoDBPersistenceLayer):
	_instance = None

	# def __new__(cls, *_, **__):
	# 	if cls._instance is None:
	# 		cls._instance = super(DynamoDBPersistenceLayer, cls).__new__(cls)
	# 	return cls._instance


def mapArgsToKwargs(function: Callable, args: tuple, kwargs: dict) -> tuple[tuple, dict]:
	"""
	assumption: valid invocation signature
	"""
	argsList: list = list(args)
	kwargsNew: dict = kwargs.copy()

	signature: inspect.Signature = inspect.signature(function)
	defaultKwargs: dict = {k: v.default for k, v in signature.parameters.items() if v.default is not inspect.Parameter.empty}
	funcArgs: list = [arg for arg, _ in signature.parameters.items()]

	for funcArg in funcArgs:
		if argsList:
			kwargsNew[funcArg] = argsList.pop(0)

	for defaultKwargKey in defaultKwargs:
		if defaultKwargKey not in kwargsNew:
			kwargsNew[defaultKwargKey] = defaultKwargs[defaultKwargKey]
	return (), kwargsNew


def retryFunction(func: Callable, args: tuple = (), kwargs: dict = {}, maxRetries: int = 5, sleepFactor: float = 1) -> any:
	retry: int = 0
	while retry < maxRetries:
		time.sleep(retry * sleepFactor)
		try:
			result: any = func(*args, **kwargs)
			return result
		except Exception as e:
			exception = e
			retry += 1
			message: str = f'function "{func.__qualname__}" with args {args} and kwargs {kwargs} failed. retry count at {retry}: '
			print(message + str(e))
	time.sleep(0.1)
	raise exception


region_config = Config(
	region_name='us-east-1',
)
dynamo: DynamoDBPersistenceLayer = DynamoDBPersistenceLayer(table_name=idempotentTableName, boto_config = region_config)

def idempotentFunction(cacheKey: str, expectedDurationSeconds: int = 5, expiresAfterSeconds: int = 60 * 5):
	def decorator(function: Callable):
		@functools.wraps(function)
		def wrapper(*args, **kwargs):
			argsNew, kwargsNew = mapArgsToKwargs(function, args, kwargs)

			config: IdempotencyConfig = IdempotencyConfig(expires_after_seconds=expiresAfterSeconds)

			dynamo.configure(config=config, function_name=function.__qualname__)
			hashKey: str = dynamo._get_hashed_idempotency_key(kwargsNew[cacheKey].__dict__)
			functionToInvoke: Callable = idempotent_function(function, data_keyword_argument=cacheKey, persistence_store=dynamo, config=config)

			print(f'invoking idempotent function {function.__module__}.{function.__qualname__}, {hashKey} , with data {kwargsNew[cacheKey].__dict__}')
			# returnValue = helper.retryFunction(functionToInvoke, args=argsNew, kwargs=kwargsNew, sleepFactor=expectedDurationSeconds)
			returnValue = functionToInvoke(*argsNew, **kwargsNew)
			return returnValue

		return wrapper

	return decorator


class Side(str, Enum):
	BUY: str = 'buy'
	SELL: str = 'sell'


@dataclass
class TradeOrder:
	isin: str
	quantity: int
	side: Side
	limitPrice: float = None
	stopPrice: float = None

	def __str__(self) -> str:
		return f'isin: {self.isin}, quantity {self.quantity}, type {self.side.value}, optional limit/stop price {self.limitPrice}/{self.stopPrice}'


t1 = TradeOrder(isin='US7140461093', quantity=200, side=Side(Side.BUY))
t2 = TradeOrder(isin='US7140461091', quantity=201, side=Side(Side.BUY))


@idempotentFunction(cacheKey='tradeOrder')
def myFancyFunction(tradeOrder: TradeOrder):
	print(tradeOrder)
	time.sleep(5)
	return 'data'


def invoke(t: TradeOrder):
	myFancyFunction(t)


threads = [
	Thread(target=invoke, args=(t1,)),
	Thread(target=invoke, args=(t2,)),
]
for thread in threads:
	thread.start()

for thread in threads:
	thread.join()
```

### Possible Solution

Simply instantiating the client as part of the persistance layer itself, instead of re-instantiating the client upon access of the property mitigates the problem to a large part. In particular, the chance of collisions is reduced greatly. In particular, if thread instantiation itself is sequentially done, then this already mitigates the problem entirely. 
This is a patch for DynamoDBPersistanceLayer:

 ```
Index: venv/Lib/site-packages/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/venv/Lib/site-packages/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py b/venv/Lib/site-packages/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py
--- a/venv/Lib/site-packages/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py	
+++ b/venv/Lib/site-packages/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py	(date 1668465118921)
@@ -95,6 +95,7 @@
         self.status_attr = status_attr
         self.data_attr = data_attr
         self.validation_key_attr = validation_key_attr
+        self.ddb_resource = self._boto3_session.resource("dynamodb", config=self._boto_config)
         super(DynamoDBPersistenceLayer, self).__init__()
 
     @property
@@ -105,8 +106,8 @@
         """
         if self._table:
             return self._table
-        ddb_resource = self._boto3_session.resource("dynamodb", config=self._boto_config)
-        self._table = ddb_resource.Table(self.table_name)
+
+        self._table = self.ddb_resource.Table(self.table_name)
         return self._table
 
     @table.setter
```

### Steps to Reproduce

(see code snippet above - "minimal breaking example" ) 

### AWS Lambda Powertools for Python version

latest

### AWS Lambda function runtime

3.9

### Packaging format used

PyPi

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by mploski**
**Issue number**: https://github.com/awslabs/aws-lambda-powertools-python/issues/1718

## Summary

PR modifies idempotency feature to make sure decorating user function with idempotency decorator is thread safe
### Changes

> Please provide a summary of what's being changed

PR modifies idempotency feature to use low level boto3 dynamodb client which is thread safe. It also adds other features/fixes as:

- End2end test to ensure function decorated with idempotency features can be used in multiple parallel threads
- Fix pylance type checking in an if condition
- Add [type](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/dynamodb.html) to DynamoDB conditions/items in Idempotency utility and functional tests
- Fix some other small stylistic errors

### User experience

> Please share what the user experience looks like before and after this change

User experience stays the same, but user doesn't need to worry about thread safety
## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [ ] Changes are documented
* [ ] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
error (must be from resource client -> low-level client):

> >               raise NoRegionError()
> E               botocore.exceptions.NoRegionError: You must specify a region.

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py [hunk 1]
```diff
@@ -79,13 +81,14 @@ def __init__(
 
         self._boto_config = boto_config or Config()
         self._boto3_session = boto3_session or boto3.session.Session()
+        self._client = self._boto3_session.client("dynamodb", config=self._boto_config)
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:84]: @heitorlessa sent a feedback here: https://github.com/awslabs/aws-lambda-powertools-python/pull/1899#issuecomment-1418791593

File: tests/e2e/idempotency/test_idempotency_dynamodb.py [hunk 2]
```diff
@@ -22,12 +22,16 @@ def parallel_execution_handler_fn_arn(infrastructure: dict) -> str:
...
+    return infrastructure.get("ParallelFunctionsHandlerArn", "")
+
+
 @pytest.fixture
 def idempotency_table_name(infrastructure: dict) -> str:
     return infrastructure.get("DynamoDBTable", "")
 
 
-@pytest.mark.xdist_group(name="idempotency")
 def test_ttl_caching_expiration_idempotency(ttl_cache_expiration_handler_fn_arn: str):
```
Comment (leandrodamascena) [inline tests/e2e/idempotency/test_idempotency_dynamodb.py]: Recently @rubenfonseca worked to improve the e2e tests in this PR: https://github.com/awslabs/aws-lambda-powertools-python/pull/1861
I think you need to rebase/merge your local branch with the latest updates and add this decorator `@pytest.mark.xdist_group(name="XX")` before test function signature.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hello @mploski! Another great job improving the Idempotency utility! I made some comments and I think we need to address them before starting a review to merge this.

Could you please check the comments and resolve them?

Thank you very much.

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py (hunk 1 above)
Comment (mploski) [reply aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:84]: Fixed breaking test by passing region name explicitly

**Comment by mploski** [comment]
> 

Hey @leandrodamascena, thx for looking into this :-) I fixed last pytest error and reverted back lines related to pytest tests grouping. Please re-review if you have some time.

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py [hunk 3]
```diff
@@ -136,36 +123,39 @@ def _item_to_data_record(self, item: Dict[str, Any]) -> DataRecord:
...
+            self.status_attr: {"S": data_record.status},
         }
 
         if data_record.in_progress_expiry_timestamp is not None:
-            item[self.in_progress_expiry_attr] = data_record.in_progress_expiry_timestamp
+            item[self.in_progress_expiry_attr] = {"N": str(data_record.in_progress_expiry_timestamp)}
 
-        if self.payload_validation_enabled:
-            item[self.validation_key_attr] = data_record.payload_hash
+        if self.payload_validation_enabled and data_record.payload_hash:
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:157]: I'm curious about that. Did you add this condition to make the code more resilient, or did you discover a bug in the depths of that code? :laughing:

File: tests/e2e/idempotency/test_idempotency_dynamodb.py [hunk 4]
```diff
@@ -97,3 +102,30 @@ def test_parallel_execution_idempotency(parallel_execution_handler_fn_arn: str):
     # THEN
     assert "Execution already in progress with idempotency key" in error_idempotency_execution_response
     assert "Task timed out after" in timeout_execution_response
+
+
+@pytest.mark.xdist_group(name="idempotency")
+def test_parallel_functions_execution_idempotency(parallel_functions_handler_fn_arn: str):
```
Comment (leandrodamascena) [inline tests/e2e/idempotency/test_idempotency_dynamodb.py]: I think this name might be confusing for someone looking at the code. The same goes for the `test_parallel_execution_idempotency` test. We should take this opportunity to rename both. Let me think of good names for this or you can suggest.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
I just made a suggestion to make the code more readable for other people. But I tested it locally and everything is working as expected. Tomorrow I will do another review and I think we are very close to a merger.
Thanks a lot @mploski, your codes always bring me great insights.

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py (hunk 3 above)
Comment (mploski) [reply aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:157]: That was raised to me by pylance type checking engine in my VSCode ( typing rocks!). Payload_hash is an optional string in data record object, so we could have a problem if we would have set payload_validation_enabled to true and wouldn't pass hash.  This fix makes pylance and myself happy :D

File: tests/e2e/idempotency/test_idempotency_dynamodb.py (hunk 4 above)
Comment (mploski) [reply tests/e2e/idempotency/test_idempotency_dynamodb.py]: happy to hear your suggestion, I will try to come up with better proposal this evening also

Comment (mploski) [reply tests/e2e/idempotency/test_idempotency_dynamodb.py]: @leandrodamascena Renamed the `test_parallel_functions_execution_idempotency` to `test_idempotent_function_thread_safety` ( + renamed handlers accordingly). Let me know what do you think. Have no idea how to rename `test_parallel_execution_idempotency` so any suggestion welcomed :-)

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py (hunk 3 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py:157]: Ohh I see! Great catch, dude!

File: tests/e2e/idempotency/test_idempotency_dynamodb.py (hunk 4 above)
Comment (leandrodamascena) [reply tests/e2e/idempotency/test_idempotency_dynamodb.py]: Thank you, @mploski! Everything is fine now. With this change, clean up what you're testing and avoid confusion.

I don't have a good name for the other function right now, so we can merge this PR and I'll add in my backlog to rename this function in next releases.

**Review by leandrodamascena — APPROVED** [review]
I added a slight change to the documentation. Thanks so much for the effort to make this utility even more resilient, @mploski!

If you don't have any considerations, we are ready to merge!

image

**Comment by mploski** [comment]
> I added a slight change to the documentation. Thanks so much for the effort to make this utility even more resilient, @mploski!
> 
> If you don't have any considerations, we are ready to merge!
> 
> image

@leandrodamascena thanks for reviewing it! :-) Let's ship it

**Pull request merged**